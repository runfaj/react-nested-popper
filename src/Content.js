import React from 'react';
import PropTypes from 'prop-types';
import { createPopper } from '@popperjs/core';
import _ from 'lodash';

import Queue from './Queue';
import Portal from './Portal';

export default class Content extends React.Component {
  static className = 'react-nested-popper_Content';

  popperEl = null;
  portalEl = null;
  popperInstance = null;

  constructor(props) {
    super(props);

    // we add a unique id so the queue can easily find instances as needed
    this.id = _.uniqueId();
  }

  componentDidMount() {
    // if show was initially enabled, init popperjs immediately
    if (this.props._show) {
      this.initPopperInstance();
    }
  }

  componentDidUpdate() {
    // watch prop changes to see if we need to init or update the popperjs instance
    // if not show, we kill the instance and clean up refs
    if (this.props._show) {
      if (!this.popperInstance) {
        this.initPopperInstance();
      } else {
        this.popperInstance.setOptions(this.popperOptions);
      }
    } else {
      this.destroyPopperInstance();
    }
  }

  componentWillUnmount() {
    this.destroyPopperInstance();
  }

  /*  called if queue wants to hide this popper */
  componentWillDestroy() {
    this.props._onComponentWillDestroy();
  }

  /*  init popperjs */
  initPopperInstance() {
    // only init if we actually have a target and popper element to bind to
    if (this.popperEl && this.props._targetRef && !this.popperInstance) {
      this.popperInstance = createPopper(this.props._targetRef, this.popperEl, this.props.popperOptions);
      // push this instance to the queue so the queue can manage it
      Queue._push(this, this.props._queue);
      document.addEventListener('click', this.onOutsideClick);
    }
  }

  /*  destroy popperjs */
  destroyPopperInstance() {
    if (this.popperInstance) {
      document.removeEventListener('click', this.onOutsideClick);
      this.popperInstance.destroy();
      this.popperInstance = null;
      // remove this instance from the queue
      Queue._removeBy(this);
    }
    this.popperEl = null;
    this.portalEl = null;
  }

  /*  sets the ref for the popperjs content so it can be bound. Also provide to user if they want it */
  setPopperRef = (el) => {
    this.popperEl = el;
    if (this.props.innerRef) {
      this.props.innerRef(el);
    }
  }

  /*  track the portal ref, used for outside click checking */
  setPortalRef = (el) => {
    this.portalEl = el;
  }

  /*  any click outside the portal or popper should trigger an outside click */
  onOutsideClick = (e) => {
    if (this.portalEl) {
      if (!this.portalEl.contains(e.target)) {
        this.props._onOutsideClick(this, e);
      }
    } else if (!this.popperEl || !this.popperEl.contains(e.target)) {
      this.props._onOutsideClick(this, e);
    }
  }

  render() {
    if (!this.props._targetRef || !this.props._show) {
      return null;
    }

    const popper = (
      <div
        className={this.props.className}
        ref={this.setPopperRef}
        onClick={this.props.onClick}
      >
        {this.props.children}
        {this.props.includeArrow && (
          <div className={this.props.arrowClassName} data-popper-arrow />
        )}
      </div>
    );

    if (this.props._usePortal) {
      return (
        <Portal
          portalRoot={this.props._portalRoot}
          className={this.props.portalClassName}
          innerRef={this.setPortalRef}
        >
          {popper}
        </Portal>
      );
    }

    return popper;
  }
}

// see docs for definitions of each external prop here
Content.propTypes = {
  // external
  arrowClassName: PropTypes.string,
  className: PropTypes.string,
  includeArrow: PropTypes.bool,
  innerRef: PropTypes.func,
  onClick: PropTypes.func,
  popperOptions: PropTypes.object,
  portalClassName: PropTypes.string,
  
  // internal - these can't be required otherwise will show error to end user
  _onComponentWillDestroy: PropTypes.func,
  _onOutsideClick: PropTypes.func,
  _portalRoot: PropTypes.element,
  _queue: PropTypes.string,
  _show: PropTypes.bool,
  _targetRef: PropTypes.any,
  _usePortal: PropTypes.bool,
};
Content.defaultProps = {
  arrowClassName: '',
  className: '',
  includeArrow: false,
  innerRef: (el) => {},
  onClick: (e) => {},
  popperOptions: {},
  portalClassName: '',
  
  _onComponentWillDestroy: () => {},
  _onOutsideClick: (instance, e) => {},
  _portalRoot: null,
  _queue: 'global',
  _show: false,
  _targetRef: null,
  _usePortal: true,
};
