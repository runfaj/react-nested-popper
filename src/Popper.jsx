import React from 'react';

const Popper = ({
  target = null,
  content = null,
}) => (
  <div data-test="test">
    <div>{target}</div>
    <div>{content}</div>
  </div>
);

export default Popper;
