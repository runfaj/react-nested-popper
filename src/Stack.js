export const StackStore = {};

export class Stack {
  debug = false;
  pushing = {};

  constructor(stack, debug = false) {
    if (stack) {
      this.getStack(stack);
    }
    this.debug = debug;
  }

  /* ******** public stack methods ******** */
  /*  gets existing stack by stack name, or creates new empty stack */
  getStack = (stack = 'global') => {
    let newStack = StackStore[stack];
    if (newStack) {
      return newStack;
    }

    newStack = [];
    StackStore[stack] = newStack;
    this._log('new stack:', stack);
    return newStack;
  }

  /*  destroy one or all stacks. "true" will kill all, otherwise will destroy by given names */
  destroyStack = (stacks = true) => {
    if (stacks === true) {
      Object.keys(StackStore).forEach(this.destroyStack);
      return;
    }

    if (typeof stacks === 'string') {
      stacks = [stacks];
    }

    // stacks are provided in a stack (not confusing at all). Only destroy the highest stack
    // in the list that contains items
    for (let q = stacks.length - 1; q >= 0; q -= 1) {
      const stackName = stacks[q];
      const currStack = this.getStack(stackName);
      if (currStack.length > 0) {
        // if we are pushing to a stack, we don't want to remove anything. Because technically click outside and target
        // click fire together, we have to cancel the destroy that happens in that scenario
        if (this.pushing[stackName]) {
          return;
        }

        // loop through all objects in given stack and tell them to kill themselves (that sounds so evil)
        // we let the component lifecycles handle the actual removing so we aren't breaking react flows
        for (let i = currStack.length - 1; i >= 0; i -= 1) {
          const instance = currStack[i];
          if (instance.componentWillDestroy && !instance.isDestroying) {
            instance.componentWillDestroy();
          }
        }

        this._log('destroy stack:', stackName);

        break;
      }
    }
  }

  /* ******** internal methods ******** */
  /*  kind of like a pop method, this will destroy the last instance on a given stack
      See notes in destroyStack for the multiple stacks
  */
  _destroyLast = (instance, stacks = 'global') => {
    if (typeof stacks === 'string') {
      stacks = [stacks];
    }

    // stacks are provided in a stack (not confusing at all). Only destroy the highest stack
    // in the list that contains items
    for (let q = stacks.length - 1; q >= 0; q -= 1) {
      const stackName = stacks[q];
      const currStack = this.getStack(stackName);
      if (currStack.length > 0) {
        if (this.pushing[stackName]) {
          return;
        }
        
        // get the last instance. We also have to provide an instance to this method because multiple instances could
        // trigger this method on outside click at the same time, so we want to only allow the highest visible instance
        // to destroy itself
        const lowestInstance = currStack[currStack.length - 1];
        if (lowestInstance.id !== instance.id) {
          return;
        }

        this._log('_destroyLast: should destroy:', instance.id, instance.isDestroying);
        if (instance.componentWillDestroy && !instance.isDestroying) {
          instance.componentWillDestroy();
        }

        break;
      }
    }
  }

  /*  push an instance onto a given stack, making it the highest visible in the stack */
  _push = (instance, stacks = 'global') => {
    if (typeof stacks === 'string') {
      stacks = [stacks];
    }

    stacks.forEach(stack => {
      // because onClickOutside can be called at the same time as a push, depending on context, we flag if a push is happening
      this.pushing[stack] = true;
      this.getStack(stack).push(instance);
      this._log('push:', stack, instance.id);
      this._clearPushing(stack);
    });
  }

  /*  reflow is a special case for "auto" where we provide the indexes of the stack parents. This allows us to
      to close sibling contents. Returns new list of stacks to push content onto.
  */
  _reflow = (targetRef, stacks) => {
    const outputs = [];
    Object.keys(stacks).forEach(parentStack => {
      const foundIdx = stacks[parentStack];
      const parentIsLast = StackStore[parentStack].length - 1 === foundIdx;
      outputs.push(parentStack);

      // move all open siblings to new stack
      if (!parentIsLast) {
        const siblings = StackStore[parentStack].splice(foundIdx + 1);
        const firstSibling = siblings[0];
        const newStack = '_content' + firstSibling.id;
        this.getStack(newStack).push(...siblings);
        for (let i = siblings.length - 1; i >= 0; i -= 1) {
          const instance = siblings[i];
          instance.determinedStack = [newStack];
          if (instance.onOutsideClick) {
            instance.onOutsideClick(targetRef);
          }
        }
      }

      /*
        if someone comes up with the case - we may need to re-address this. Currently, if no poppers have the outside close
        and we have:
           p
          / \
          1 3
          | |
          2 4
        Where the following order opens: p 1 2 3 4,
        then our stacks become
        p 1 2
        3 4
        However, we might want it like
        p 1 3
        1 2
        3 4
        First attempts of this caused weird mis-matches, so we'll need a real-life use case
      */
    });

    return outputs;
  }

  /*  this will manually splice an instance out of the stack. This is called by the component lifecycle, so this
      effectively is a cleanup method
  */
  _removeBy = (instance) => {
    const { id } = instance;

    Object.keys(StackStore).forEach(key => {
      const items = StackStore[key];
      const idx = items.findIndex(item => item.id === id);
      if (idx > -1) {
        items.splice(idx, 1);
        this._log('removeBy:', key, instance.id);
      }
    });
  }

  /*  after we push an instance, we need to clear the pushing state. This is debounced so we only clear after all
      possible pushing has completed
  */
  _clearPushing = (stack = 'global') => {
    setTimeout(() => {
      if (this.pushing[stack]) {
        delete this.pushing[stack];
      }
    }, 250);
  }

  /*  logging function. Just enable the debug flag on the stack */
  _log = (...msg) => {
    if (this.debug) {
      console.log('Stack:', ...msg, StackStore);
    }
  }

  /*  debugging helper function to see stack store */
  _getStackStore = () => {
    if (this.debug) {
      return StackStore;
    }
    return null;
  }

  /* given a target, returns all content parent stacks that contain target */
  _getTargetParents = (targetRef) => {
    /*
        Test not written for this method, as it does some dom checking that is very hard to accurately test.
        Essentially, we want to take an input target element and check all the stacks (which contain content
        elements) and see if the target directly belongs to any of those content elements. If so, we track
        the stack name and index it was found.

        So, if I had this:
          StackStore = {
            global: [Content1, Content2, Content3],
            blah: [Content2, Content3],
          }
        And my target was found nested under Content2, my returned value would be
          {
            global: 1,
            blah: 0
          }
        
        This output is mostly useless by itself, but directly used in the _reflow method to close contents
        that are above the found indexes.
    */

    if (!targetRef) {
      return null;
    }

    // generate list of stacks and the index each was found at
    const foundList = {};
    let foundOne = false;
    Object.keys(StackStore).forEach(key => {
      const items = StackStore[key];
      for (let i = 0; i < items.length; i += 1) {
        const content = items[i];
        if (content.popperEl.contains(targetRef)) {
          foundOne = true;
          foundList[key] = i;
          break;
        }
      }
    });

    if (!foundOne) {
      return null;
    }

    return foundList;
  }
}

// we don't export the actual stack util, since we want to always init the package with a global stack available
const StackUtil = new Stack();

export default StackUtil;
