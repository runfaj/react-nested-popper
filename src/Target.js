import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';

export default class Target extends React.Component {
  static className = 'react-nested-popper_Target';


  /*  tracks ref so popperjs can be bound to it. Also provide to user if they want it.
      We wouldn't even need this component, but we need to guarantee a ref to the target,
      so we enforce this component.
  */
  setRef(el) {
    if (this.props.innerRef) {
      this.props.innerRef(el);
    }
    if (this.props._targetRef) {
      this.props._targetRef(el);
    }
  }

  render() {
    const rest = _.omit(this.props, _.keys(Target.propTypes));

    return (
      <div
        className={this.props.className}
        ref={el => this.setRef(el)}
        onClick={this.props._onClick}
        {...rest}
      >
        {this.props.children}
      </div>
    );
  }
}

// see docs for definitions of each external prop here
Target.propTypes = {
  // external props
  className: PropTypes.string,
  innerRef: PropTypes.func,

  // internal props - these can't be required unless you want it to show to end user
  _onClick: PropTypes.func,
  _targetRef: PropTypes.func,
};
Target.defaultProps = {
  className: '',
  innerRef: (el) => {},

  _onClick: (e) => {},
  _targetRef: (el) => {},
};
