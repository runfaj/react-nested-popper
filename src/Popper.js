import React from 'react';
import PropTypes from 'prop-types';

import Target from './Target';
import Content from './Content';
import Queue from './Queue';

export default class Popper extends React.Component {
  static className = 'react-nested-popper_Popper';

  constructor(props) {
    super(props);

    this.validChildren = this.checkChildren(props.children);
    this.isManaged = this.checkIsManaged(props);

    this.state = {
      show: !!props.initiallyOpen, // managed popper uses state, otherwise user provides props.show
      targetRef: null, // we do this in the state instead of regular this assignment because we need to re-trigger a render when this updates
    };
  }

  shouldComponentUpdate(nextProps) {
    this.validChildren = this.checkChildren(nextProps.children);
    this.isManaged = this.checkIsManaged(nextProps);
    // don't bother updating if user changed the children to be invalid
    return this.validChildren;
  }

  /* get the show state */
  get show() {
    return this.isManaged ? this.state.show : this.props.show;
  }

  /*  Checks children to find the target component instance
      We only grab the first one found. Any other ones after that will be invalid, so we ignore them.
  */
  getTargetComponent(children) {
    let comp = null;
    React.Children.forEach(children, child => {
      if (!comp && child.type && child.type.className === Target.className) {
        comp = child;
      }
    });
    return comp;
  }

  /*  Checks children to find the content component instance
      We only grab the first one found. Any other ones after that will be invalid, so we ignore them.
  */
  getContentComponent(children) {
    let comp = null;
    React.Children.forEach(children, child => {
      if (!comp && child.type && child.type.className === Content.className) {
        comp = child;
      }
    });
    return comp;
  }

  /*  logs an error if children are invalid and returns valid state */
  checkChildren(children) {
    const valid = this.getTargetComponent(children) && this.getContentComponent(children);

    if (!valid) {
      console.error('Popper must contain one Target component and one Content component.');
    }

    return valid;
  }

  /* see if popper is managed or controlled based on given props */
  checkIsManaged(props) {
    return typeof props.show === 'undefined' || props.show === null;
  }

  /*  our target and content children must be internally modified in order for our queueing and binding
      to work properly. This adds those internal props to the children
  */
  modifyChildren(children) {
    return React.Children.map(children, child => {
      let newProps = {};
      if (child.type.className === Target.className) {
        newProps = {
          _onClick: this.onTargetClick,
          _targetRef: this.setTargetRef,
        };
      } else if (child.type.className === Content.className) {
        newProps = {
          _portalRoot: this.props.portalRoot,
          _usePortal: this.props.usePortal,
          _show: this.show,
          _targetRef: this.state.targetRef,
          _onComponentWillDestroy: this.onContentWillDestroy,
          _onOutsideClick: this.onContentOutsideClick,
          _queue: this.props.groupName,
        };
      }
      return React.cloneElement(child, newProps);
    });
  }

  /*  tracks the element ref to the target. This is used to bind popperjs to the target, but also to trigger a
      re-render when this is available so the content can initialize properly
  */
  setTargetRef = (el) => {
    if (!el) {
      return;
    }
    if (el === this.state.targetRef) {
      return;
    }
    this.setState({
      targetRef: el,
    });
  }

  /* 'nuff said */
  closeContent = () => {
    this.setState({
      show: false,
    });
  }

  /* if managed, this will set the show state accordingly */
  onTargetClick = () => {
    if (this.isManaged) {
      this.setState(prevState => ({
        show: this.props.targetToggle ? !prevState.show : true,
      }));
    }
  }

  /*  when we click outside the content, we need to handle that event accordingly */
  onContentOutsideClick = (contentInstance, e) => {
    // if the clicked thing contains our target, we don't bother doing anything, since onTargetClick handles that
    if (!this.state.targetRef || !this.state.targetRef.contains(e.target)) {
      // convenience function in case the user needs it for whatever reason
      if (this.props.onOutsideClick) {
        this.props.onOutsideClick(contentInstance, e);
      }
      // if we're managing the content, we get to control the queue, based on given props
      if (this.isManaged && this.props.closeOnOutsideClick) {
        switch (this.props.outsideClickType) {
          case 'group': Queue.destroyQueue(this.props.groupName); break;
          case 'all': Queue.destroyQueue(true); break;
          default: Queue._destroyLast(contentInstance, this.props.groupName);
        }
      }
    }
  }

  /*  if something this going to trigger the content to be destroyed (like killing it in the queue), we need to
      sync that here, or provide that event to the user
  */
  onContentWillDestroy = () => {
    if (this.isManaged) {
      this.closeContent();
    } else if (this.props.onPopperWillClose) {
      this.props.onPopperWillClose();
    }
  }

  render() {
    if (!this.validChildren) {
      return null;
    }

    const { children } = this.props;

    return (
      <>
        {this.modifyChildren(this.getTargetComponent(children))}
        {this.modifyChildren(this.getContentComponent(children))}
      </>
    );
  }
}

// see docs for definitions of each prop here
Popper.propTypes = {
  closeOnOutsideClick: PropTypes.bool,
  groupName: PropTypes.string,
  initiallyOpen: PropTypes.bool,
  onOutsideClick: PropTypes.func,
  onPopperWillClose: PropTypes.func,
  outsideClickType: PropTypes.oneOf(['default', 'group', 'all']),
  portalRoot: PropTypes.element,
  show: PropTypes.bool,
  targetToggle: PropTypes.bool,
  usePortal: PropTypes.bool,
};
Popper.defaultProps = {
  closeOnOutsideClick: false,
  groupName: 'global',
  initiallyOpen: false,
  onOutsideClick: (e) => {},
  onPopperWillClose: () => {},
  outsideClickType: 'default',
  portalRoot: null,
  show: null,
  targetToggle: false,
  usePortal: true,
};
