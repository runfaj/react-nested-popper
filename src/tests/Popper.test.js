import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import Target from '../Target';
import Content from '../Content';
import Popper from '../Popper';

import Stack from '../Stack';
jest.mock('../Stack');

describe('Popper component', () => {
  console.error = jest.fn();

  let targetEl, contentEl;
  beforeEach(() => {
    targetEl = <div id="target">react-nested-popper</div>;
    contentEl = <div id="content">react-nested-popper</div>;
  });

  it('renders with valid children', () => {
    const component = renderer.create(
      <Popper>
        <Target>{targetEl}</Target>
        <Content>{contentEl}</Content>
      </Popper>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('doesn\'t render with invalid children', () => {
    const component = renderer.create(
      <Popper>
        <Target>{targetEl}</Target>
      </Popper>
    );
    expect(component.toJSON()).toBeNull();

    const component2 = renderer.create(
      <Popper>
        <Content>{contentEl}</Content>
      </Popper>
    );
    expect(component2.toJSON()).toBeNull();
  });

  it('managed vs controlled set via show prop', () => {
    const managed = mount(
      <Popper>
        <Target>{targetEl}</Target>
        <Content>{contentEl}</Content>
      </Popper>
    );
    
    const controlled = mount(
      <Popper show>
        <Target>{targetEl}</Target>
        <Content>{contentEl}</Content>
      </Popper>
    );

    expect(managed.instance().isManaged).toBe(true);
    expect(controlled.instance().isManaged).toBe(false);
  });

  it('target ref is set appropriately', () => {
    const component = mount(
      <Popper>
        <Target>{targetEl}</Target>
        <Content>{contentEl}</Content>
      </Popper>
    );

    expect(component.instance().state.targetRef).not.toBeNull();
  });

  it('onTargetClick checks', () => {
    const nonToggle = mount(
      <Popper>
        <Target>{targetEl}</Target>
        <Content>{contentEl}</Content>
      </Popper>
    );
    const toggle = mount(
      <Popper targetToggle>
        <Target>{targetEl}</Target>
        <Content>{contentEl}</Content>
      </Popper>
    );

    const onControlledClick = jest.fn();
    const controlled = mount(
      <Popper show onTargetClick={onControlledClick}>
        <Target>{targetEl}</Target>
        <Content>{contentEl}</Content>
      </Popper>
    );
    
    // non-toggled state becomes show and only show
    nonToggle.instance().onTargetClick();
    expect(nonToggle.instance().state.show).toBe(true);
    nonToggle.instance().onTargetClick();
    expect(nonToggle.instance().state.show).toBe(true);
    expect(onControlledClick).toHaveBeenCalledTimes(0);

    // toggled instance, state should toggle
    toggle.instance().onTargetClick();
    expect(toggle.instance().state.show).toBe(true);
    toggle.instance().onTargetClick();
    expect(toggle.instance().state.show).toBe(false);
    expect(onControlledClick).toHaveBeenCalledTimes(0);

    // controlled instance, onTargetClick should fire
    controlled.instance().onTargetClick('event');
    expect(onControlledClick).toHaveBeenCalledWith('event');
  });

  it('closeContent checks', () => {
    const component = mount(
      <Popper>
        <Target>{targetEl}</Target>
        <Content>{contentEl}</Content>
      </Popper>
    );

    component.setState({ show: true });
    component.instance().closeContent();
    expect(component.instance().state.show).toBe(false);
  });

  it('onContentWillDestroy checks', () => {
    const onPopperWillClose = jest.fn();
    const managed = mount(
      <Popper onPopperWillClose={onPopperWillClose}>
        <Target>{targetEl}</Target>
        <Content>{contentEl}</Content>
      </Popper>
    );
    const controlled = mount(
      <Popper show onPopperWillClose={onPopperWillClose}>
        <Target>{targetEl}</Target>
        <Content>{contentEl}</Content>
      </Popper>
    );

    managed.setState({ show: true });
    managed.instance().onContentWillDestroy();
    expect(managed.instance().state.show).toBe(false);
    expect(onPopperWillClose).toHaveBeenCalledTimes(0);

    controlled.instance().onContentWillDestroy();
    expect(onPopperWillClose).toHaveBeenCalledTimes(1);
  });

  it('onContentOutsideClick checks', () => {
    const contentInstance = { foo: 'bar' };
    const e = {
      target: document.createElement('div'),
    };
    const onOutsideClick = jest.fn();

    const noClose = mount(
      <Popper onOutsideClick={onOutsideClick}>
        <Target>{targetEl}</Target>
        <Content>{contentEl}</Content>
      </Popper>
    );
    const closeDefault = mount(
      <Popper
        onOutsideClick={onOutsideClick}
        closeOnOutsideClick
        outsideClickType="default"
        groupName="global"
      >
        <Target>{targetEl}</Target>
        <Content>{contentEl}</Content>
      </Popper>
    );
    const closeAll = mount(
      <Popper
        onOutsideClick={onOutsideClick}
        closeOnOutsideClick
        outsideClickType="all"
        groupName="global"
      >
        <Target>{targetEl}</Target>
        <Content>{contentEl}</Content>
      </Popper>
    );
    const closeGroup = mount(
      <Popper
        onOutsideClick={onOutsideClick}
        closeOnOutsideClick
        outsideClickType="group"
        groupName="global"
      >
        <Target>{targetEl}</Target>
        <Content>{contentEl}</Content>
      </Popper>
    );

    noClose.instance().onContentOutsideClick(contentInstance, e);
    expect(onOutsideClick).toHaveBeenCalledWith(contentInstance, e);
    expect(Stack._destroyLast).not.toHaveBeenCalled();

    closeDefault.instance().onContentOutsideClick(contentInstance, e);
    expect(Stack._destroyLast).toHaveBeenCalledWith(contentInstance, 'global');

    closeAll.instance().onContentOutsideClick(contentInstance, e);
    expect(Stack.destroyStack).toHaveBeenCalledWith(true);

    closeGroup.instance().onContentOutsideClick(contentInstance, e);
    expect(Stack.destroyStack).toHaveBeenCalledWith('global');
  });
});
