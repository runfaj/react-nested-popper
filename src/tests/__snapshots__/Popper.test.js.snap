// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popper component renders with valid children 1`] = `
<div
  className=""
  onClick={[Function]}
>
  <div
    id="target"
  >
    react-nested-popper
  </div>
</div>
`;
